### PR TITLE
C-408 Send OAuth read-only response

### PR DESCRIPTION
**Pull request by nicoback2**
### Description
Form JWT containing user's profile information to send back to the redirect URI.

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### CODE REVIEW

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 1]
```diff
@@ -1,27 +1,119 @@
...
-  const { search } = useLocation()
-  const { scope, api_key, state, redirect } = queryString.parse(search)
+  const { search, hash } = useLocation()
+  const isRedirectValidRef = useRef<undefined | boolean>(undefined)
+  const { scope, state, redirect_uri } = queryString.parse(search)
+  const parsedRedirectUri = useMemo(() => {
+    if (redirect_uri && typeof redirect_uri === 'string') {
+      let res: URL
+      try {
+        res = new URL(redirect_uri)
```
Comment (sliptype) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:23]: Can we just return here?
```suggestion
        return new URL(redirect_uri)
```

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 2]
```diff
@@ -1,27 +1,119 @@
...
+    parsedRedirectUri == null ? null : parsedRedirectUri.host
+  const { token } = queryString.parse(hash)
   const account = useSelector(getAccountUser)
   const isLoggedIn = Boolean(account)
-  const hasMetaMask = !!(window as any).web3
   const [email, setEmail] = useState('')
   const [password, setPassword] = useState('')
   const [submitError, setSubmitError] = useState<string | null>(null)
 
+  useEffect(() => {
```
Comment (sliptype) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:40]: What's the reasoning behind making `isRedirectValid` a ref?

You could use `useMemo` here I believe:

```suggestion
  const isRedirectValid = useMemo(() => {
```

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 3]
```diff
@@ -49,30 +142,51 @@ export const OAuthLoginPage = () => {
...
-      // window.location.reload()
-    } catch (err) {
-      console.error({ err })
+  const authAndRedirect = async () => {
+    const jwt = await formOAuthResponse()
+    if (jwt == null) {
+      return
+    }
+    const statePart = state != null ? `state=${state}&` : ''
+    const fragment = `#${statePart}token=${jwt}`
```
Comment (sliptype) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:151]: Why put the token in the hash rather than as a query param?

**Review by sliptype — APPROVED** [review]
This is looking great! The redirect validation is definitely interesting bc we aren't registering the url, but good enough for now

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 2 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:40]: No real reason! I'll change it so it's more grokkable

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--oauth-response

**Pull request merged**